### PR TITLE
PNAS citations shown via popup windows

### PR DESCRIPTION
**Pull request by zhongh**
screen shot 2016-04-10 at 9 22 14 pm

### CODE REVIEW

**Comment by zhongh** [comment]
Decided to push a commit first with citation being displayed via popup window.

P.S. Having been trying time after times to figure out how a "hover" popup box could work, but it just won't. I feel this is a simpler solution, but not too bad.

**Comment by zednis** [comment]
I think we want the citation next to be copyable so users can copy&paste, as such I don't think a standard hover popup is ideal.

Have you looked at Bootstrap JS for pop-up ideas?

File: html/publications.html [hunk 1]
```diff
@@ -4,17 +4,29 @@
...
-    <script type="text/javascript" src="/vivo/themes/dco/js/facetview2/vendor/bootstrap/js/bootstrap.min.js"></script>
-    <link rel="stylesheet" href="/vivo/themes/dco/js/facetview2/vendor/jquery-ui-1.8.18.custom/jquery-ui-1.8.18.custom.css">
-    <script type="text/javascript" src="/vivo/themes/dco/js/facetview2/vendor/jquery-ui-1.8.18.custom/jquery-ui-1.8.18.custom.min.js"></script>
-
-    <script type="text/javascript" src="/vivo/themes/dco/js/facetview2/es.js"></script>
-    <script type="text/javascript" src="/vivo/themes/dco/js/facetview2/bootstrap2.facetview.theme.js"></script>
-    <script type="text/javascript" src="/vivo/themes/dco/js/facetview2/jquery.facetview2.js"></script>
-    <link rel="stylesheet" href="/vivo/themes/dco/js/facetview2/css/facetview.css">
-    <link rel="stylesheet" href="/vivo/themes/dco/css/browsers.css">
+    <!--<script type="text/javascript" src="/vivo/themes/dco/js/facetview2/vendor/jquery/1.7.1/jquery-1.7.1.min.js"></script>-->
```
Comment (mrpatrickwest) [inline html/publications.html]: This should not be part of the branch. Need to change this back to what it was in production. Only your local copy should have the src path changes, not what you commit.

Comment (zhongh) [inline html/publications.html]: Sure I will fix this in the next commit

**Comment by zhongh** [comment]
Issue:
@zednis @mrpatrickwest 
Unless we are going to use a simple <a> link to the citation service, any popovers usually need the JS to load the html content returned by a url like "https://dco.tw.rpi.edu/citation/11121/1015-6779-1645-7537-CC". However in attempts to implement this, errors are constantly shown regarding Cross Origin Request.

I have tried something like 

```
 $("button").click(function(){
        $.ajax({
            type: "GET",
            crossOrigin: true,
            url: "https://dco.tw.rpi.edu/citation/11121/1015-6779-1645-7537-CC",
            dataType: 'html',
            success: function(data) {
                console.log(data);
            }
        });
    });
```

or this:

```
 $("*").load("https://dco.tw.rpi.edu/citation/11121/1015-6779-1645-7537-CC");
```

or this library http://www.html5rocks.com/en/tutorials/cors/ with this:

```
var xhr = createCORSRequest('GET', "https://dco.tw.rpi.edu/citation/11121/1015-6779-1645-7537-CC");
xhr.send();
```

Same cross origin request error is always shown.

**Comment by zednis** [comment]
background info - https://en.wikipedia.org/wiki/Cross-origin_resource_sharing
- see https://www.kerstner.at/en/2015/02/enabling-cross-origin-resource-sharing-cors-for-php/
- We ran into this problem with the PHP opensearch services from the old S2S browsers.
  see how we set `header("Access-Control-Allow-Origin: *");`
  in https://github.com/tetherless-world/s2s/blob/master/opensearch/config.php
- it appears our solution for the old S2S opensearch services will not work for all browsers (e.g. firefox), so look to the first link for a solution.

**Comment by zhongh** [comment]
@zednis @mrpatrickwest Ooops minutes earlier I found that the ajax actually works when put on dco.tw.

**Comment by zednis** [comment]
Do you know why?

**Comment by zhongh** [comment]
@xgmachina said that he met this problem before and these get request will simply work when it is not on localhost, but on a server. He said that it needs to be on valid url.
I have been searching things like "localhost cross-origin request blocked" and I think this is around the reason why? @zednis

**Comment by zednis** [comment]
read the background info link from my earlier post

**Comment by zhongh** [comment]
@zednis Sure! Thanks!

**Comment by zhongh** [comment]
@zednis @mrpatrickwest I finally tried out on a plain html that in this way the popover works the way I want:

```
<!DOCTYPE html>
<head>
    <meta charset="utf-8">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.0/jquery.min.js"></script>
    <script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js"></script>
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css">

    <script type="text/javascript">
        var citation_service_url = "https://dco.tw.rpi.edu/citation/";

        var options = {
            content: 'Default popover content',
            html: true,
            placement: 'auto top',
        };

        var invokePopover = function(source) {

            escaped_id = source.id.replace("/", "\\/");

            $(source).attr("data-content", "Changed!");
            $('#' + escaped_id + '[data-toggle="popover"]').popover(options);
            $('#' + escaped_id + '[data-toggle="popover"]').popover("toggle");

            /* // This part is temporarily commented out for local testing of the popover feature.
            $.ajax({
                type: "GET",
                crossOrigin: true,
                url: citation_service_url + source.id,
                success: function(data) {
                    $(source).attr("data-content", data);
                    $('#' + escaped_id + '[data-toggle="popover"]').popover(options);
                    $('#' + escaped_id + '[data-toggle="popover"]').popover("toggle");
                }
            });
            */
        };
    </script>
</head>
<body>
        <h2>A badge:</h2>

        <a id="11121/1015-6779-1645-7537-CC" href="#" title="PNAS Citation" data-toggle="popover" data-trigger="manual" data-content="Content" onclick="invokePopover(this)">
            <img style="vertical-align: middle" src="https://dcotest.tw.rpi.edu/badges/badge?title=citation&text=PNAS&color=5C85D6">
        </a>

        <h2>Another badge:</h2>

        <a id="11121/1015-6779-1645-7537-AA" href="#" title="PNAS Citation" data-toggle="popover" data-trigger="manual" data-content="Content" onclick="invokePopover(this)">
            <img style="vertical-align: middle" src="https://dcotest.tw.rpi.edu/badges/badge?title=citation&text=PNAS&color=5C85D6">
        </a>

</body>
</html>
```

But when I put it correspondingly in the publication browser code on both my local and https://dco.tw.rpi.edu/vivo/publications, the popover won't pop and console gives error "jquery-1.7.1.min.js:2 Uncaught TypeError: Cannot read property 'replace' of undefined".

**Comment by zednis** [comment]
The citation pop-up from my recent commit:

screen shot 2016-04-13 at 9 42 17 pm

It takes a while (10+ seconds) for the pop-up to be ready as the browser invokes an AJAX request for each search result when the page loads.  The citation service takes a while to answer all 20 AJAX requests.  The pop-up will show "loading..." until the citation text has been retrieved from the citation service.

I think there are a few next steps to improve performance
- [ ] add caching to the citation service so it returns cached responses when able
- [ ] update the information in ES so the citation can be generated from the search result object (or possibly generate citations during ingest and retrieve them directly from ES)

edit: if we can generate or retrieve citation text from ES it will be very fast.

**Comment by zhongh** [comment]
@zednis Do you know why changing the browser structure to using handlebars fixed the issue?

Using information in ES will of course let the citation be generated in no time, but it seems then we are not using the citation service.

Or we can let the citation be generated when clicked on one badge, while others still be generated as the page being loaded?

**Comment by zednis** [comment]
@zhonghao1014 using HandlebarsJS itself did not fix the issue, that was just a refactor I threw in to clean-up the HTML generation.  One issue may have been that the HTML id (at least for HTML 4) must begin with a non-number alpha character [A-Za-z] and can then include numbers and a few special characters (but not slashes).  In my update I have "id_" added before the DCO-ID and replace the slash with an underscore; the result is an HTML id such as "id_11121_1015-6779-1645-7537-CC" which is legal and which jquery can successfully use in the selectors that toggle the popover and update the data-content attribute.

I am not 100% sure where you were having problems, but since I had some example source working on my end I decided to implement and commit it for DCO.

**Comment by zhongh** [comment]
👍 

I am still wondering slightly why mine won't work as mine too replace the slash in the ids and it works locally. But anyway now we have a much cleaner HTML and a working popover. Thanks!

**Comment by zhongh** [comment]
@zednis I made some trial changes to the code, haven't commit here, but put it on here: https://dco.tw.rpi.edu/vivo/publications
The popover is not preloaded all at once, but when you click it, and wait a few seconds, then it will change to result.

**Comment by zednis** [comment]
@zhonghao1014 I see one bug on the uncommitted code on dco.tw.  If you open pop-up and close it before the AJAX request returns it will be re-open on its own when the AJAX request finishes.

**Comment by mrpatrickwest** [comment]
I say :shipit: and we can create branches to fix any problems we encounter and switching to elasticsearch queries instead of sparql queries

**Comment by zednis** [comment]
I say not yet to merging, lets fix the bug I just noticed first.

**Comment by zhongh** [comment]
Let's me quickly fix the bug since this bug is related to the task.

**Comment by zhongh** [comment]
@zednis @mrpatrickwest check it now: https://dco.tw.rpi.edu/vivo/publications
- citation service request is sent per item when it is first clicked
- if the popover remains on, it updates its content when result is received
- if the popover has already been closed before the result is received, it won't popup unless you click it again.
  If good I can upload the commit here

**Comment by zednis** [comment]
@zhonghao1014 in the future please commit code before putting it into test :-)

**Comment by zhongh** [comment]
Ready? https://dco.tw.rpi.edu/vivo/publications

**Comment by zednis** [comment]
:+1: what do you think @mrpatrickwest?

**Comment by mrpatrickwest** [comment]
:+1:

**Comment by zednis** [comment]
:shipit:

**Pull request merged**